File: scripts/audit-orphans.mjs
import fs from 'fs';
import path from 'path';

const SRC_DIR = './src';
const SUPABASE_DIR = './supabase';
const DOCS_DIR = './docs';

function getAllFiles(dir, exts, fileList = []) {
  if (!fs.existsSync(dir)) return fileList;
  const files = fs.readdirSync(dir);
  for (const file of files) {
    const filePath = path.join(dir, file);
    if (fs.statSync(filePath).isDirectory()) {
      getAllFiles(filePath, exts, fileList);
    } else {
      if (exts.some(ext => filePath.endsWith(ext))) {
        fileList.push(filePath.replace(/\\/g, '/'));
      }
    }
  }
  return fileList;
}

const allCodeFiles = [
  ...getAllFiles(SRC_DIR, ['.ts', '.tsx']),
  ...getAllFiles(SUPABASE_DIR, ['.ts', '.sql'])
];

function getDocumentedFiles() {
  const docFiles = getAllFiles(DOCS_DIR, ['.md']);
  const codeRefs = new Set();
  
  for (const doc of docFiles) {
    const content = fs.readFileSync(doc, 'utf-8');
    const match = content.match(/code_refs:\s*\n([\s\S]*?)(?:^[a-z_-]+:|\n---)/m);
    if (match) {
      const lines = match[1].split('\n');
      for (const line of lines) {
        const refMatch = line.match(/^\s*-\s*(.+)$/);
        if (refMatch) {
          codeRefs.add(refMatch[1].trim());
        }
      }
    }
  }
  return Array.from(codeRefs);
}

const documented = getDocumentedFiles();

const orphans = allCodeFiles.filter(file => {
  // file looks like src/pages/Index.tsx
  for (const ref of documented) {
    // ref looks like src/pages/ ou src/pages/Index.tsx
    if (file === ref || file.startsWith(ref)) return false;
  }
  return true;
});

console.log(`Total code files: ${allCodeFiles.length}`);
console.log(`Documented references: ${documented.length}`);
console.log(`Orphan files: ${orphans.length}`);

if (!fs.existsSync('./docs/_audit')) {
  fs.mkdirSync('./docs/_audit', { recursive: true });
}
fs.writeFileSync('./docs/_audit/orphans.txt', orphans.join('\n'));
console.log('Orphan list saved to docs/_audit/orphans.txt');

File: scripts/audit_orphans.py
import os
import re

SRC_DIR = './src'
SUPABASE_DIR = './supabase'
DOCS_DIR = './docs'

def get_all_files(directory, exts):
    file_list = []
    if not os.path.exists(directory):
        return file_list
    for root, dirs, files in os.walk(directory):
        for file in files:
            if any(file.endswith(ext) for ext in exts):
                # normalize path to posix style with leading ./ removed if any, but since we start with './', let's format consistently.
                # os.walk uses backslashes on windows.
                file_path = os.path.join(root, file).replace('\\', '/')
                # remove leading './' to match the frontmatter style easily which usually starts with 'src/' or 'supabase/'
                if file_path.startswith('./'):
                    file_path = file_path[2:]
                file_list.append(file_path)
    return file_list

all_code_files = get_all_files(SRC_DIR, ['.ts', '.tsx']) + get_all_files(SUPABASE_DIR, ['.ts', '.sql'])

def get_documented_files():
    doc_files = get_all_files(DOCS_DIR, ['.md'])
    code_refs = set()
    
    for doc in doc_files:
        with open(doc, 'r', encoding='utf-8') as f:
            content = f.read()
            match = re.search(r'code_refs:\s*\n([\s\S]*?)(?:^[a-z_-]+:|\n---)', content, re.MULTILINE)
            if match:
                lines = match.group(1).split('\n')
                for line in lines:
                    ref_match = re.match(r'^\s*-\s*(.+)$', line)
                    if ref_match:
                        ref = ref_match.group(1).strip()
                        # normalize
                        if ref.startswith('./'):
                            ref = ref[2:]
                        code_refs.add(ref)
    return list(code_refs)

documented = get_documented_files()
orphans = []

for file in all_code_files:
    is_documented = False
    for ref in documented:
        if file == ref or file.startswith(ref):
            is_documented = True
            break
    if not is_documented:
        orphans.append(file)

print(f"Total code files: {len(all_code_files)}")
print(f"Documented references: {len(documented)}")
print(f"Orphan files: {len(orphans)}")

os.makedirs('./docs/_audit', exist_ok=True)
with open('./docs/_audit/orphans.txt', 'w', encoding='utf-8') as f:
    f.write('\n'.join(orphans))
print("Orphan list saved to docs/_audit/orphans.txt")
